src/dsa/rpo_falcon512/signature.rs

```diff
@@ -39,251 +42,334 @@ use super::{
-            }
-            ctx
-        }
+/// Takes the hash-to-point polynomial `c` of a message, the signature polynomial over
+/// the message `s2` and a public key polynomial and returns `true` is the signature is a valid
+/// signature for the given parameters, otherwise it returns `false`.
+fn verify_helper(c: Polynomial<FalconFelt>, s2: SignaturePoly, h: PubKeyPoly) -> bool {
+    let h_fft = h.fft();
+    let s2_fft = s2.fft();
+    let c_fft = c.fft();
 
-        /// Absorbs data into the RPO state.
-        pub fn absorb(&mut self, data: &[u8]) {
-            unsafe {
-                rpo128_absorb(
-                    self as *mut Rpo128Context,
-                    data.as_ptr() as *const c_void,
-                    data.len(),
-                )
-            }
-        }
+    // s1 = c - s2 * h
+    let s1_fft = c_fft - s2_fft.hadamard_mul(&h_fft);
+    let s1 = s1_fft.ifft();
+
+    let length_squared_s1 = s1.norm_squared();
+    let length_squared_s2 = s2.norm_squared();
+    let length_squared = length_squared_s1 + length_squared_s2;
+
+    length_squared < SIG_L2_BOUND
+}
+
+/// Checks whether a set of coefficients is a valid one for a signature polynomial.
+fn are_coefficients_valid(x: &[i16]) -> bool {
+    if x.len() != N {
+        return false;
+    }
 
-        /// Finalizes the RPO state to prepare for squeezing.
-        pub fn finalize(&mut self) {
-            unsafe { rpo128_finalize(self as *mut Rpo128Context) }
+    for &c in x {
+        if !(-2047..=2047).contains(&c) {
+            return false;
         }
     }
 
-    #[test]
-    fn test_hash_to_point() {
-        // Create a random message and transform it into a u8 vector
-        let msg_felts: Word = rand_vector::<Felt>(4).try_into().unwrap();
-        let msg_bytes = msg_felts
-            .iter()
-            .flat_map(|e| e.as_int().to_le_bytes())
-            .collect::<alloc::vec::Vec<_>>();
-
-        // Create a nonce i.e. a [u8; 40] array and pack into a [Felt; 8] array.
-        let nonce: [u8; 40] = rand_vector::<u8>(40).try_into().unwrap();
-
-        let mut buffer = [0_u8; 64];
-        for i in 0..8 {
-            buffer[8 * i] = nonce[5 * i];
-            buffer[8 * i + 1] = nonce[5 * i + 1];
-            buffer[8 * i + 2] = nonce[5 * i + 2];
-            buffer[8 * i + 3] = nonce[5 * i + 3];
-            buffer[8 * i + 4] = nonce[5 * i + 4];
+    true
+}
+
+/// Takes as input a list of integers x and returns a bytestring that encodes/compress' it.
+/// If this is not possible, it returns False.
+///
+/// For each coefficient of x:
+/// - the sign is encoded on 1 bit
+/// - the 7 lower bits are encoded naively (binary)
+/// - the high bits are encoded in unary encoding
+///
+/// This method can fail, in which case it returns None.
+///
+/// Algorithm 17 p. 47 of the specification [1].
+///
+/// [1]: https://falcon-sign.info/falcon.pdf
+fn compress_signature(x: &[i16]) -> Vec<u8> {
```

Comment:
Similar comment as above: should we move the body of this function into `Serializable::write_into()`?